### PR TITLE
🐛Cleanup ports

### LINKED ISSUE #1062: CreateInstance leaks ports on certain error conditions
/kind bug

**What steps did you take and what happened:**
[A clear and concise description of what the bug is.]
As higlighted by #1061, CreateInstance will leak ports if the specified image or flavor doesn't exist, or if glance or nova return any other error while validating them.

https://github.com/kubernetes-sigs/cluster-api-provider-openstack/blob/7735d4d714f4572a4d086a78319003fdcf109702/pkg/cloud/services/compute/instance_test.go#L668-L669

https://github.com/kubernetes-sigs/cluster-api-provider-openstack/blob/7735d4d714f4572a4d086a78319003fdcf109702/pkg/cloud/services/compute/instance_test.go#L684-L685

**What did you expect to happen:**
Ports should always be cleaned up on any error until they are attached to a server.

### PR DESCRIPTION
**Pull request by mdbooth**
**This PR is only the final commit in the series! All other commits are from #1061, which needs to merge first**

**What this PR does / why we need it**:
This fixes a leak of ports when failing to fetch image or flavor for any reason, including user error, during instance creation.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #1062

**Special notes for your reviewer**:

1. Please confirm that if this PR changes any image versions, then that's the sole change this PR makes.

**TODOs**:

- [X] Merge #1061 
- [X] Rebase so that only the final commit remains

/hold until #1061 merges

### CODE REVIEW

**Review by chrischdi — COMMENTED** [review]
The business looks way more clean :+1: 

I'd say mission accomplished :smiley: 

/lgtm

**Comment by chrischdi** [comment]
Looks like some prow / gcloud error unrelated to this PR on a first view?

**Comment by mdbooth** [comment]
> Looks like some prow / gcloud error unrelated to this PR on a first view?

Kubelet didn't come up correctly on the control plane node:
https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/kubernetes-sigs_cluster-api-provider-openstack/1063/pull-cluster-api-provider-openstack-e2e-test/1463518333197357056/artifacts/clusters/bootstrap/instances/e2e-d43a5h/cluster-e2e-d43a5h-control-plane-p2h2j/cloud-final.log

I haven't yet worked out how to read kubelet's startup logs, so it's anybody's guess why. It's a flake, though.

**Comment by chrischdi** [comment]
> > Looks like some prow / gcloud error unrelated to this PR on a first view?
> 
> Kubelet didn't come up correctly on the control plane node: https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/kubernetes-sigs_cluster-api-provider-openstack/1063/pull-cluster-api-provider-openstack-e2e-test/1463518333197357056/artifacts/clusters/bootstrap/instances/e2e-d43a5h/cluster-e2e-d43a5h-control-plane-p2h2j/cloud-final.log
> 
> I haven't yet worked out how to read kubelet's startup logs, so it's anybody's guess why. It's a flake, though.

All I can see is that the kubelet itself is not able to connect to the kube-apiserver to register himself as node.

I also can't find any logs of the static pods :-(

**Comment by mdbooth** [comment]
From CAPO controller logs:
```
I1125 15:22:55.469152       1 recorder.go:103] events "msg"="Warning"  "message"="Failed to delete security group k8s-cluster-e2e-158vcy-cluster-e2e-158vcy-secgroup-worker with id 94947127-7579-40e3-960d-98e1b600b423: Expected HTTP response code [] when accessing [DELETE http://10.0.3.15:9696/v2.0/security-groups/94947127-7579-40e3-960d-98e1b600b423], but got 409 instead\n{\"NeutronError\": {\"type\": \"SecurityGroupInUse\", \"message\": \"Security Group 94947127-7579-40e3-960d-98e1b600b423 in use.\", \"detail\": \"\"}}" "object"={"kind":"OpenStackCluster","namespace":"e2e-158vcy","name":"cluster-e2e-158vcy","uid":"19281045-6a3e-4d05-bb9f-72fe2868a2a9","apiVersion":"infrastructure.cluster.x-k8s.io/v1beta1","resourceVersion":"6621"} "reason"="FailedDeleteSecurityGroup"
E1125 15:22:55.470519       1 controller.go:317] controller/openstackcluster "msg"="Reconciler error" "error"="failed to delete security groups: Expected HTTP response code [] when accessing [DELETE http://10.0.3.15:9696/v2.0/security-groups/94947127-7579-40e3-960d-98e1b600b423], but got 409 instead\n{\"NeutronError\": {\"type\": \"SecurityGroupInUse\", \"message\": \"Security Group 94947127-7579-40e3-960d-98e1b600b423 in use.\", \"detail\": \"\"}}" "name"="cluster-e2e-158vcy" "namespace"="e2e-158vcy" "reconciler group"="infrastructure.cluster.x-k8s.io" "reconciler kind"="OpenStackCluster" 
```

I can't see any obvious reason for this in either the CAPO or the devstack logs. If I had to put my money somewhere, I would guess this is a neutron bug. Everything else seems to have been deleted AFAICT.

**Comment by mdbooth** [comment]
Same issue: failure to delete security group:

```
I1126 13:28:32.913722       1 recorder.go:103] events "msg"="Warning"  "message"="Failed to delete security group k8s-cluster-e2e-3elhda-cluster-e2e-3elhda-secgroup-worker with id adde714f-5834-4e02-83e5-92ba02e555d6: Expected HTTP response code [] when accessing [DELETE http://10.0.3.15:9696/v2.0/security-groups/adde714f-5834-4e02-83e5-92ba02e555d6], but got 409 instead\n{\"NeutronError\": {\"type\": \"SecurityGroupInUse\", \"message\": \"Security Group adde714f-5834-4e02-83e5-92ba02e555d6 in use.\", \"detail\": \"\"}}" "object"={"kind":"OpenStackCluster","namespace":"e2e-3elhda","name":"cluster-e2e-3elhda","uid":"5c537fb3-a12e-47a4-bc95-92cc01478386","apiVersion":"infrastructure.cluster.x-k8s.io/v1beta1","resourceVersion":"6334"} "reason"="FailedDeleteSecurityGroup"
```

It's the same security group and the same test in both cases, so this is starting to look deterministic.

Possibly related: I notice that several changes landed in neutron stable/victoria on 24th November, which is when this change seems to have started failing.

**Comment by mdbooth** [comment]
It's a bug introduced by this patch! `client.CreateServer() returns `*ServerExt, error`. The port cleanup function doesn't cleanup if server != nil, but `CreateServer` *never* returns nil for the server, even if there was an error 🤦‍♂️

Consequently the failed create doesn't cleanup ports, and the security group is still in use.

I've probably made this mistake systematically in a few places. I'll update.

**Comment by mdbooth** [comment]
Fixed it 🎉

Still need to merge #1061 first, which is under review with @macaptain.

File: pkg/cloud/services/compute/instance.go [hunk 1]
```diff
@@ -206,10 +206,20 @@ func (s *Service) createInstance(eventObject runtime.Object, clusterName string,
 		})
 	}
 
-	if instanceSpec.Subnet != "" && accessIPv4 == "" {
+	var server *ServerExt
+
+	// Ensure we delete the ports we created if we haven't created the server.
+	defer func() {
```
Comment (chrischdi) [inline pkg/cloud/services/compute/instance.go:212]: should the defer maybe also be above line 184?

If creating multiple ports and a port fails we may want to cleanup the already created ones too?

**Review by chrischdi — COMMENTED** [review]
The other PR was merged.

One last idea: do we want to move the defer for deleting ports above the ports creation?

File: pkg/cloud/services/compute/instance.go (hunk 1 above)
Comment (mdbooth) [reply pkg/cloud/services/compute/instance.go:212]: I can also add a unit test for that.

Comment (mdbooth) [reply pkg/cloud/services/compute/instance.go]: I've squashed this and added a new unit test.

**Comment by macaptain** [comment]
Nice!

/lgtm

**Pull request merged**